stan/math/prim/mat/fun/qr_thin_Q.hpp

```diff
@@ -0,0 +1,30 @@
+#ifndef STAN_MATH_PRIM_MAT_FUN_QR_THIN_Q_HPP
+#define STAN_MATH_PRIM_MAT_FUN_QR_THIN_Q_HPP
+
+#include <stan/math/prim/arr/err/check_nonzero_size.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/err/check_greater_or_equal.hpp>
+#include <Eigen/QR>
+#include <algorithm>
+
+namespace stan {
+namespace math {
+
+template <typename T>
+Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic> qr_thin_Q(
```

Comment:
Can this just not all live in qr_Q/qr_R instead of being a new function? It's the same interface in Eigen.

Reply:
I should have never introduced the fat QR decomposition, but I didn't recognize how to do the thin version at the time. But if we change or remove the fat QR decomposition now, it would break some working Stan programs.

Reply:
Oh I see, so qr_Q is gonna return the full Q always and that's a different shape. I guess this does need to be separate if we don't want things to break. Seems like a shame though, but I guess this is fine for now.

Maybe when tuples come around, and there's a [Q, R] = qr(m) function, that one would be automatically thin?